Final series lessons - pyproject, code of conduct & license files (#181)
pyOpenSci/python-package-guide: tutorials/add-license-coc.md

# Add a LICENSE & CODE_OF_CONDUCT to your Python package

In the [previous lesson](add-readme) you:

<i class="fa-solid fa-circle-check" style="color: #703c87;"></i> Created a basic `README.md` file for your scientific Python package

<i class="fa-solid fa-circle-check" style="color: #703c87;"></i> Learned about the core components that are useful to have in a README file.

:::{admonition} Learning objectives
:class: tip

In this lesson you will learn:

1. How to select and add a `LICENSE` file to your package repository with a focus on the GitHub interface.
2. How to add a `CODE_OF_CONDUCT` file to your package repository.
3. How you can use the Contributors Covenant website to add generic language as a starting place for your `CODE_OF_CONDUCT`.
:::

## Add a LICENSE file to your project directory

A license contains legal language about how users can use and reuse your software. You should include a LICENSE file in your project directory that specifies the license that you choose for your package.

> We suggest that you use a permissive license that accommodates the other most commonly used licenses in the scientific Python ecosystem (MIT[^mit] and BSD-3[^bsd3]).

Reviewer: Maybe I am a bit later in this discussion, and maybe it is something that you have already discussed. Sorry if I am causing any noise.

I think maybe thi paragraph coul be a bit different l, maybe something like:

You should use any open source library accepted by OSI (link) that fits better your purpose. If are not sure about what license you should use, we recommend you to use one of those commonly used licenses in the scientific Python ecosystem BSD3 or MIT.

---

Not sure if OSI would the best reference, but of there is any other better it could be changed to that as well

Author: perfect! generally folks have been suggesting choosealicense.com so i've added this

> If you are unsure, use MIT given it's the generally recommended
> license on [choosealicense.com](https://choosealicense.com/).

let me know what you think and thanks for the input!! not too late at all. i'm just starting to review and integrate feedback now.

Reviewer: That looks good to me :)

Author: yay! ok resolving. thank you Ivan.

:::{admonition} Licenses for the scientific Python ecosystem
[We discuss licenses for the scientific Python ecosystem in more detail here in our guidebook.](permissive-license)
:::

### Where should the LICENSE file live & how do you add it?

Your `LICENSE` file should be placed at the root of your package's repository.

There are several ways to add a license file:

> 1. When you create a new repository on GitHub, it will ask you if you wish to add a `LICENSE` file at that time. It will also create the file for you.

Reviewer 2: ```suggestion
1. When you create a new repository on GitHub, it will ask you if you wish to add a `LICENSE` file at that time. If you select yes, it will create the file for you.
```

2. You can add a license through the GitHub gui following the [<i class="fa-brands fa-github"></i> instructions here](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/adding-a-license-to-a-repository).
3. You can add the file manually like we are doing in this lesson.

:::{tip}
If you completed the past lessons including

1. [Making your code installable](1-installable-code.md) and
2. [publishing your package to PyPI](publish-pypi.md)

> then you already have a **LICENSE** file containing text for the MIT license in your Python package. Thus you can skip to the next section of this tutorials which walks you through adding a CODE_OF_CONDUCT.

Reviewer 2: ```suggestion
then you already have a **LICENSE** file containing text for the MIT license in your Python package. Thus you can skip to the next section of this tutorial which walks you through adding a CODE_OF_CONDUCT.
```

If you don't yet have a **LICENSE** file in your directory, then continue reading.
:::

### How to add a LICENSE to your existing GitHub repository

If you don't already have a LICENSE file, and you are not yet using a platform such as GitHub or GitLab, then you can create a license file by

> 1. Create a new file called LICENSE. If you are using shell you can use:

Reviewer 2: ```suggestion
1. Create a new file called LICENSE. If you are using a shell you can use:
```

:::

::::::{dropdown} Instructions for adding a license files within the GitHub interface
:color: primary

:::::{tab-set}

::::{tab-item} Add license: new GitHub repository

When you create a new GitHub repository you can add a license
through the GitHub interface.

:::{figure-md} github-new-repo

<img src="../images/tutorials/github-new-repo.png" alt="Screenshot of the create new repository interface that GitHub provides. The elements of this are the owner and repository name for the new repo. Below that you can add a description of the repo. Below that you can set it to be public or private. At the bottom of the interface there is an Add a README checkbox where it will add a blank readme file for you. At the very bottom there is a line to add a .gitignore file and another to choose a license." width="500px">

Image showing the GitHub interface that allows you to add a LICENSE and README file when you create a new repository.
:::
::::

::::{tab-item} Add License: Existing GitHub repository

> If you already have a GitHub repo for your package, then you can add a LICENSE using the GitHub interface by adding a new file to the repo.

Reviewer 2: This is the only place that "reop" is used as an abbreviation (and the image alt-text which mixes them). Should probably lengthen to repository. But also should maybe add to whatever glossary we are building as this is a common short-hand, but I don't think it is a recognizable abbreviation to those that don't already know what a repository is.

Author: ok excellent. i've fixed this. i can't find it in a find and replace anywhere else on this page but i'll keep looking for it. i did find it in our publish to pypi lesson and fixed there as well! great point

* Follow the instructions to select and add a license to your repository on the [GitHub LICENSE page](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/adding-a-license-to-a-repository) .
* Once you have added your LICENSE file, be sure to sync your git local repository with the repository on GitHub.com. This means running `git pull` to update your local branch.

:::{figure-md} view-license
<img src="../images/tutorials/view-license-github.png" alt="sdfsdfsd nasdfjsdf" width="500px">

> You can also view an overview of the license on GitHub if you view it in the GitHub interface.

Reviewer 2: ```suggestion
You can also view a summary of the license on its GitHub landing page.```

:::
::::

:::::
::::::

(add-coc)=
## Add a CODE_OF_CONDUCT file to your repo

> Now that you have added a LICENSE to your project, you are ready to add a file called `CODE_OF_CONDUCT.md` to your package directory. A `CODE_OF_CONDUCT` file is critical to supporting your community as it

Reviewer 3: I would add that the COC file to the root of the repository. For LICENSE and CODE_OF_CONDUCT files, if they're at the root, GitHub makes them super easy to find as tabs next to the README

eg ⏬ 
<img width="642" alt="Screenshot 2024-02-13 at 6 08 26 PM" src="https://github.com/pyOpenSci/python-package-guide/assets/54685329/f6433505-5378-459b-adb2-cc80afe3b889">

Reviewer 3: I've seen a lot of different locations that are all valid, but if we can be opinionated, root is the way to go!

Author: perfect - i've added a similar image and updated the text! we should definitely be opinionated about these things to move towards some consistency. this is great.

[![Contributor Covenant](https://img.shields.io/badge/Contributor%20Covenant-2.1-4baaaa.svg)](#)

### Add your CODE_OF_CONDUCT file

* Add a `CODE_OF_CONDUCT.md` file to your repository if it doesn't
already exist.
* Visit the [contributor covenant website](https://www.contributor-covenant.org/) and add [the markdown version of their code of conduct](https://www.contributor-covenant.org/version/2/1/code_of_conduct/code_of_conduct.md) to your `CODE_OF_CONDUCT.md` file.

:::{admonition} Additional Code of Conduct resources
:class: note

* [<i class="fa-brands fa-github"></i> Guide: `CODE_OF_CONDUCT.md` files](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/adding-a-code-of-conduct-to-your-project)
* [pyOpenSci package guide `CODE_OF_CONDUCT.md` overview](https://www.pyopensci.org/python-package-guide/documentation/repository-files/code-of-conduct-file.html)

:::

## <i class="fa-solid fa-hands-bubbles"></i> Wrap up

In this lesson and the [last lesson](add-readme), you have added a:

* `README` file;
* `LICENSE` file and a
* `CODE_OF_CONDUCT` file.

> These are core files needed for every scientific Python package. In the upcoming

Reviewer: I think it needs some kind of explanation about the term "core" here, because in my mind core would be the files needed to run the package. But I don't have too much suggestions though ... maybe it could be something like "key files to describe and distribute your package"

Author: i hear you. i've modified it to say this:

> These are fundamental files needed for every scientific Python package 
> repository. These files help users understand how to use your package and interact 
> with package maintainers.

<i class="fa-solid fa-circle-check" style="color: #703c87;"></i> Learned about the core components that are useful to have in a README file.

:::{admonition} Learning objectives
:class: tip

In this lesson you will learn:

1. How to select and add a `LICENSE` file to your package repository with a focus on the GitHub interface.
2. How to add a `CODE_OF_CONDUCT` file to your package repository.
3. How you can use the Contributors Covenant website to add generic language as a starting place for your `CODE_OF_CONDUCT`.
:::

## Add a LICENSE file to your project directory

A license contains legal language about how users can use and reuse your software. You should include a LICENSE file in your project directory that specifies the license that you choose for your package.

We suggest that you use a permissive license that accommodates the other most commonly used licenses in the scientific Python ecosystem (MIT[^mit] and BSD-3[^bsd3]). If you are unsure, use MIT given it's the generally recommended
license on [choosealicense.com](https://choosealicense.com/).

:::{admonition} Licenses for the scientific Python ecosystem

> [We discuss licenses for the scientific Python ecosystem in more detail here in our guidebook.](permissive-license)

Author: good catch ! fixed.

:::

> ### Where should the LICENSE file live & how do you add it?

Author: i like this. thank you! i feel like explaining what a license is first however is important.  I did reorganize things however but kept that overview first!

> ### Where should the LICENSE file live & how do you add it?

Reviewer 4: totally yes, "what is a license," "select a license" "implement a license" sounds like a good order

> ### Where should the LICENSE file live & how do you add it?

Author: ok great. i'm going through this one last time, i'll merge and then those 6 pr's you opened - i don't want to lose that work. so if you can open them against main here rather than in my fork that would be super awesome!

Image showing the GitHub interface that allows you to add a LICENSE and README file when you create a new repository.
:::
::::

::::{tab-item} Add License: Existing GitHub repository

If you already have a GitHub repository for your package, then you can add a LICENSE using the GitHub interface by adding a new file to the repo.

- Follow the instructions to select and add a license to your repository on the [GitHub LICENSE page](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/adding-a-license-to-a-repository) .
- Once you have added your LICENSE file, be sure to sync your git local repository with the repository on GitHub.com. This means running `git pull` to update your local branch.

:::{figure-md} view-license

> <img src="../images/tutorials/view-license-github.png" alt="sdfsdfsd nasdfjsdf" width="500px">

Reviewer 4: this image isn't found! and also the alt text might need some work lol. 

I can help write it but idk what the image is at the moment :)

Author: @sneakers-the-rat i'm not working on this until tomorrow but i had to comment as i'm laughing hard now 😆 

i am not sure how i missed that. it could be that because the image didn't get ported over from another branch, it didn't render at all and i missed it when looking at the rendered pages. i generally create dummy alt text when i suspect an image might change a few times and then update right before the thing is merged. but this is an absolute error on my part. thank you for catching.

Author: ok this has been fixed. i've been manually checking out each file from a ginormous PR i had originally with all of the lessons together. and i just missed this image - because it was in a tab. but it's rendering now and i've fixed the alt as well! great catch!

<i class="fa-solid fa-circle-check" style="color: #703c87;"></i> Learned about the core components that are useful to have in a README file.

:::{admonition} Learning objectives
:class: tip

In this lesson you will learn:

1. How to select and add a `LICENSE` file to your package repository with a focus on the GitHub interface.
2. How to add a `CODE_OF_CONDUCT` file to your package repository.
3. How you can use the Contributors Covenant website to add generic language as a starting place for your `CODE_OF_CONDUCT`.
:::

## What is a license?

A license contains legal language about how users can use and reuse your software. To set the LICENSE for your project, you:

1. create LICENSE file in your project directory that specifies the license that you choose for your package and
2. reference that file in your pyproject.toml data where metadata are set.

By adding the LICENSE file to your pyproject.toml file, the LICENSE will be included in your package's metadata which is used to populate your package's PyPI landing page. The LICENSE is also used in your GitHub repository's landing page interface.

### What license should you use?

> We suggest that you use a permissive license that accommodates the other most commonly used licenses in the scientific Python ecosystem (MIT[^mit] and BSD-3[^bsd3]). If you are unsure, use MIT given it's the generally recommended

Reviewer 4: I know that this is a guide for beginners, and probably not for this PR, but i also feel like we would be sorta remiss in not at least talking about why copyleft licenses exist and why they are important, particularly given the historical moment. We are pretty strongly selecting for audience, and also steering newbies towards "make sure your code can be gobbled up by corporations easily" without providing the alternative. I'll make a separate PR to the main repo about this, but I think it's important to cultivate a new generation that understands the importance of the free part of free and open source software

Author: this sounds great! we have a license guide type of page in the guidebook as well and this type of content might also be really useful there and we could link to it from a tutorial. but yes a separate pr would be great about copyleft and what licenses mean in terms of corporate / for profit vs open use / FOSS
